### ISSUE TITLE
All indices are considered outdated after a "borg serve" crash

### ISSUE THREAD

**Issue by wzyboy** [borgbackup/borg#4766]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes.

## Is this a BUG / ISSUE report or a QUESTION?

It's probably a bug.

#### Your borg version (borg -V).

Client: 1.1.2
Server: 1.1.5

#### Operating system (distribution) and version.

Client: CentOS 7
Server: Arch Linux

#### Hardware / network configuration, and filesystems used.

Client: Ext4
Server: NFS
Network: SSH (from US West to US East)

#### How much data is handled by borg?

The repo is 142 GiB in size, with 42 archives. Each archive contains tons of small files. (it's a backup repo of a CI server, and there are a lot of node_modules...)

#### Full borg commandline that lead to the problem (leave away excludes and passwords)

```
('borg', 'prune', 'example.org:/nfs/borgbackup/foobar', '--keep-within', '3H', '--keep-hourly', '24', '--keep-daily', '7', '--keep-weekly', '4', '--keep-monthly', '6', '--keep-yearly', '10', '--prefix', '{hostname}-', '--remote-path', 'borg', '--info', '--stats')
```
(I'm using Borgmatic)

## Describe the problem you're observing.

I'm using Borg to do daily backups from a US West client to a US East server. It worked without issues for over a year (a oldest archive in the repo is from 2018).

I have a health check service to make sure the daily backups succeed. For three days in a row, I have been receiving failure alerts. I took a look at both the client and server, and found out that the server memory was too low, so OOMs were triggered and "borg serve" processes were killed, which resulted in failed backups.

So I upgraded the server memory and retried the backup from the client. I noticed, however, this backup took way longer than usual. I checked the log and saw that the borg client was trying to rebuild the indices:

```
Sep 24 22:24:11 borgmatic[23956]: Synchronizing chunks cache...
Sep 24 22:24:11 borgmatic[23956]: Archives: 42, w/ cached Idx: 0, w/ outdated Idx: 0, w/o cached Idx: 42.
Sep 24 22:24:11 borgmatic[23956]: Fetching and building archive index for foobar-2019-09-22T00:25:46 ...
Sep 24 22:24:46 borgmatic[23956]: Merging into master chunks index ...
Sep 24 22:24:46 borgmatic[23956]: Fetching and building archive index for foobar-2019-09-13T00:13:01 ...
Sep 24 22:25:24 borgmatic[23956]: Merging into master chunks index ...
Sep 24 22:25:24 borgmatic[23956]: Fetching and building archive index for foobar-2019-04-30T00:06:26 ...
Sep 24 22:25:50 borgmatic[23956]: Merging into master chunks index ...
Sep 24 22:25:50 borgmatic[23956]: Fetching and building archive index for foobar-2019-09-19T01:27:59 ...
Sep 24 22:43:07 borgmatic[23956]: Merging into master chunks index ...
Sep 24 22:43:07 borgmatic[23956]: Fetching and building archive index for foobar-2019-08-28T00:32:11 ...
Sep 24 22:57:38 borgmatic[23956]: Merging into master chunks index ...
[...]
```

The building process went on for hours before I got impatient and killed the borg client. After restarting the job, the building process continued:

```
Sep 25 04:48:14 borgmatic[30428]: Synchronizing chunks cache...
Sep 25 04:48:14 borgmatic[30428]: Archives: 42, w/ cached Idx: 23, w/ outdated Idx: 0, w/o cached Idx: 19.
Sep 25 04:48:14 borgmatic[30428]: Fetching and building archive index for foobar-2019-09-10T01:37:57 ...
Sep 25 04:48:49 borgmatic[30428]: Merging into master chunks index ...
Sep 25 04:48:49 borgmatic[30428]: Fetching and building archive index for foobar-2019-09-01T01:15:51 ...
Sep 25 04:49:22 borgmatic[30428]: Merging into master chunks index ...
Sep 25 04:49:22 borgmatic[30428]: Reading cached archive chunk index for foobar-2019-08-22T01:15:47 ...
Sep 25 04:49:22 borgmatic[30428]: Merging into master chunks index ...
Sep 25 04:49:22 borgmatic[30428]: Fetching and building archive index for foobar-2019-09-21T01:21:10 ...
Sep 25 04:49:53 borgmatic[30428]: Merging into master chunks index ...
Sep 25 04:49:54 borgmatic[30428]: Reading cached archive chunk index for foobar-2019-09-06T00:28:13 ...
Sep 25 04:49:54 borgmatic[30428]: Merging into master chunks index ...
Sep 25 04:49:54 borgmatic[30428]: Fetching and building archive index for foobar-2019-02-28T00:11:05 ...
[...]
```

The client and the server communicates over a not-very-fast SSH connection, and the server storage is on a throttled NFS mount. So this building process has been going on for more than 6 hours...

After reading [this issue](https://github.com/borgbackup/borg/issues/4740), I understand that the indices are good for future backups. However, I am wondering why borg thinks all the indices are outdated, as pointed out by these two lines:

```
Sep 24 22:24:11 borgmatic[23956]: Archives: 42, w/ cached Idx: 0, w/ outdated Idx: 0, w/o cached Idx: 42.
Sep 25 04:48:14 borgmatic[30428]: Archives: 42, w/ cached Idx: 23, w/ outdated Idx: 0, w/o cached Idx: 19.
```

The repo is not accessed from any other clients. Only the client reads from and writes to it. IMHO there is a bug somewhere, which causes Borg to consider all chunk indices are "outdated" while they are not.

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

If I stop the client and restart it, the index building process continues. Does this count?

#### Include any warning/errors/backtraces from the system logs

The server was killed because of OOM, but it's not Borg's fault. :-)

**Comment by ThomasWaldmann** [comment]
If your client is really still running 1.1.2, you should urgently upgrade to a newer 1.1.x release - there were lots of fixes since then.

Besides the packages that might be available from/for your linux distribution, there are also the fat binary releases on our github releases page. They include everything needed, you only need a recent enough glibc.

**Comment by ThomasWaldmann** [comment]
The first time borg syncs/rebuilds the chunk index, it fetches a lot of metadata for all files for each archive in the repo. It then caches a per-archive chunks index (so it won't have to fetch / compute that per-archive chunks index again).

From all the per-archive chunks indexes, it computes the overall chunks index (this index maps chunkid -> size, csize, refcount).

If it says it had 0 cached indexes, it means this ran for the first time (or the cache has been erased).

Overall, this is a quite heavy operation, but having a correct overall chunks index is essential for correct operation. borg needs to know what chunks it has in the repo and how often each of them is referenced (the latter is needed to decide when to really delete a chunk).

**Comment by ThomasWaldmann** [comment]
If you want to help, try to reproduce your issue with relatively recent borg versions, preferably the latest release under controlled conditions (e.g. run both locally and use ssh over localhost). Then kill -9 the borg serve process and see if you can reproduce that way.

Theoretically, nothing bad should happen, as a suddenly dying borg serve process does not commit the repo. So when using the repo the next time, all data after the previous commit will be considered invalid and removed again.

On the client side, the client also won't commit its current state and also revert back to the last valid indexes/caches.

**Comment by wzyboy** [comment]
Thank you for you detailed explanation on how chunk indices work, though I figured most of it myself :-)

In fact, I am using the fat binary on the client side. (Borg was not available in the distro's official repo) I even wrote an [Ansible Playbook](https://github.com/wzyboy/borgbackup-playbook) for this task. But I did not bump the fat binary version after setting up the client since then. I should do it now!

**Comment by wzyboy** [comment]
```
Sep 25 09:12:34 borgmatic[30428]: Done.
```

It took 10+ hours to rebuild the indices of a 142 GiB repo (42 archives). The cache is 3 GiB in size, not very big. I blame SSH and NFS for this slowness.

I would update the client version and try to reproduce this.

**Comment by wzyboy** [comment]
I updated the client version to 1.1.10 and tried to reproduce this by killing borg server mid-backup. The client reported "Connection closed by remote host" as expected. After restarting the client, there was no more "chunk indices outdated" messages. I repeated this three times and the results were the same.

So I assume at least with 1.1.10, the issue does not exist anymore.

Thank you @ThomasWaldmann !
